File: backend/app/routers/location_reopen.py
"""Admin review queue for location reopen requests. See
docs/API_CONTRACTS.md "Location reopen requests" and
app/models/location_reopen_request.py.

Submission itself (`POST /locations/{id}/reopen-requests`, owner-only)
lives on `app/routers/locations.py` — nested under the location it's
about, same as `POST /locations/{id}/managers`. This router is only the
admin-facing half (list/get/approve/reject), mirroring `/claim`'s shape
(`app/routers/claim.py`) at its own top-level path rather than under
`/admin`, matching `/claim` and `/reports` (neither of which live under
`/admin` either, despite being admin-reviewed).
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies.auth import CurrentUser, get_current_user, require_admin
from app.dependencies.db import get_db
from app.dependencies.pagination import Pagination, pagination_params
from app.schemas.location_reopen import (
    ReopenRequestApproveRequest,
    ReopenRequestListResponse,
    ReopenRequestOut,
    ReopenRequestRejectRequest,
    ReopenRequestStatus,
)
from app.services import location_reopen_service

router = APIRouter(prefix="/location-reopen-requests", tags=["location-reopen-requests"])


@router.get("", response_model=ReopenRequestListResponse)
async def list_reopen_requests(
    status_filter: ReopenRequestStatus = Query(default="pending_review", alias="status"),
    pagination: Pagination = Depends(pagination_params),
    db: AsyncSession = Depends(get_db),
    _admin: CurrentUser = Depends(require_admin),
) -> ReopenRequestListResponse:
    """Auth: admin only. `status` defaults to `pending_review`; oldest first."""
    return await location_reopen_service.list_reopen_requests(db, pagination, status_filter)


@router.get("/{request_id}", response_model=ReopenRequestOut)
async def get_reopen_request(
    request_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
) -> ReopenRequestOut:
    """Auth: the requesting owner (their own request) or admin (any)."""
    request = await location_reopen_service.get_reopen_request(db, request_id, current_user)
    return location_reopen_service.to_reopen_request_out(request)


@router.post("/{request_id}/approve", response_model=ReopenRequestOut)
async def approve_reopen_request(
    request_id: int,
    body: ReopenRequestApproveRequest,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_admin),
) -> ReopenRequestOut:
    """Auth: admin. Real side effect: the location's status flips to
    `active` (location_reopen_service.approve_reopen_request)."""
    request = await location_reopen_service.approve_reopen_request(
        db, request_id, current_user.cognito_sub, body.reviewer_notes
    )
    return location_reopen_service.to_reopen_request_out(request)


@router.post("/{request_id}/reject", response_model=ReopenRequestOut)
async def reject_reopen_request(
    request_id: int,
    body: ReopenRequestRejectRequest,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_admin),
) -> ReopenRequestOut:
    """Auth: admin. `reviewer_notes` is required to reject, same as
    `POST /claim/{id}/reject`."""
    request = await location_reopen_service.reject_reopen_request(
        db, request_id, current_user.cognito_sub, body.reviewer_notes
    )
    return location_reopen_service.to_reopen_request_out(request)

File: frontend/src/lib/api/locationReopen.ts
// Typed client for location reopen requests — docs/API_CONTRACTS.md
// "Location reopen requests (`/locations/{id}/reopen-requests`,
// `/location-reopen-requests`)". Mirrors @/lib/api/claim.ts's shape.
import { apiFetch, toQueryString } from "./client";
import type { PaginatedResponse, PaginationParams } from "@/types/common";
import type {
  ApproveReopenRequestInput,
  CreateReopenRequestInput,
  RejectReopenRequestInput,
  ReopenRequestQueueItem,
  ReopenRequestResponse,
  ReopenRequestStatus,
} from "@/types/locationReopen";

/**
 * POST /locations/{id}/reopen-requests — auth: owner (must own the parent
 * brand), only for a location that is currently `closed_pending_reopen`
 * (409 otherwise, or if one is already pending).
 */
export async function submitReopenRequest(
  locationId: number,
  input: CreateReopenRequestInput,
  accessToken: string
): Promise<ReopenRequestResponse> {
  return apiFetch<ReopenRequestResponse>(
    `/locations/${locationId}/reopen-requests`,
    { method: "POST", body: JSON.stringify(input) },
    { accessToken }
  );
}

/**
 * GET /location-reopen-requests — auth: admin. Queue, oldest pending
 * first; `status` defaults to `pending_review` server-side.
 */
export async function listReopenRequests(
  params: PaginationParams & { status?: ReopenRequestStatus },
  accessToken: string
): Promise<PaginatedResponse<ReopenRequestQueueItem>> {
  const query = toQueryString({
    status: params.status,
    page: params.page,
    page_size: params.page_size,
  });
  return apiFetch<PaginatedResponse<ReopenRequestQueueItem>>(
    `/location-reopen-requests${query}`,
    { method: "GET" },
    { accessToken }
  );
}

/** GET /location-reopen-requests/{id} — auth: the requesting owner (own request) or admin (any). */
export async function getReopenRequestById(
  id: number,
  accessToken: string
): Promise<ReopenRequestResponse> {
  return apiFetch<ReopenRequestResponse>(
    `/location-reopen-requests/${id}`,
    { method: "GET" },
    { accessToken }
  );
}

/** POST /location-reopen-requests/{id}/approve — auth: admin. Flips the location back to active. */
export async function approveReopenRequest(
  id: number,
  input: ApproveReopenRequestInput,
  accessToken: string
): Promise<ReopenRequestResponse> {
  return apiFetch<ReopenRequestResponse>(
    `/location-reopen-requests/${id}/approve`,
    { method: "POST", body: JSON.stringify(input) },
    { accessToken }
  );
}

/** POST /location-reopen-requests/{id}/reject — auth: admin. reviewer_notes required. */
export async function rejectReopenRequest(
  id: number,
  input: RejectReopenRequestInput,
  accessToken: string
): Promise<ReopenRequestResponse> {
  return apiFetch<ReopenRequestResponse>(
    `/location-reopen-requests/${id}/reject`,
    { method: "POST", body: JSON.stringify(input) },
    { accessToken }
  );
}
